#!/usr/bin/env bash
set -eu -o pipefail
export PS4='+ [${BASH_SOURCE[0]##*/}:${LINENO}${FUNCNAME[0]:+:${FUNCNAME[0]}}] '
# set -x

# XXX Apply this patch after installing eos-s3server RPM.
#
# Default s3server configuration is optimized for hardware setups.
# Unless this patch is applied, s3server will consume more memory than
# is available on CI VM and will get killed by OOM-killer, failing the
# CI job; see http://gitlab.mero.colo.seagate.com/mero/hare/issues/330
{
    cat <<'EOF'
--- s3config.yaml.orig	2020-03-28 12:26:27.406026351 +0000
+++ s3config.yaml	2020-03-28 22:01:24.841026351 +0000
@@ -1,11 +1,11 @@
 S3Config_Sections: [S3_SERVER_CONFIG, S3_AUTH_CONFIG, S3_CLOVIS_CONFIG, S3_THIRDPARTY_CONFIG]
 S3_SERVER_CONFIG:                                     # Section for S3 Server
    S3_DAEMON_WORKING_DIR: "/var/mero/"                # S3 Daemon will run in the specified directory
-   S3_DAEMON_DO_REDIRECTION: 1                          # Do redirection of stderr, stdout and stdin to /dev/null
+   S3_DAEMON_DO_REDIRECTION: 0                          # Do redirection of stderr, stdout and stdin to /dev/null
    S3_LOG_DIR: "/var/log/seagate/s3"                    # S3 Log directory, for output to stderr have value as "" and set S3_DAEMON_DO_REDIRECTION to 0
    S3_LOG_MODE: INFO                                    # logging levels, levels can be DEBUG, INFO, WARN, ERROR, FATAL, default is INFO
    S3_LOG_FILE_MAX_SIZE: 100                            # Maximum log file size in MB, default is 100 MB
-   S3_LOG_ENABLE_BUFFERING: true                        # DEBUG, INFO & WARN logs are buffered if buffering is enabled. ERROR and FATAL logs are always flushed. Default is true.
+   S3_LOG_ENABLE_BUFFERING: false                        # DEBUG, INFO & WARN logs are buffered if buffering is enabled. ERROR and FATAL logs are always flushed. Default is true.
    S3_ENABLE_AUTH_SSL: false                             # Enable ssl communication between s3 server and auth server
    S3_REUSEPORT: true                                   # Enable reusing s3 server port
    S3_MERO_HTTP_REUSEPORT: true                         # Enable reusing mero http server port
@@ -36,7 +36,7 @@
    S3_ENABLE_MURMURHASH_OID: false                      # Enable OID generation using Murmur Hash Alg. Default is to have unique OID generated by mero helper library.
    S3_RETRY_INTERVAL_MILLISEC: 500                      # Retry interval in milliseconds, total retry time = retry_count * retry_interval (RETRY1: 500, RETRY2: 1000, RETRY3: 1500)
    S3_CLIENT_REQ_READ_TIMEOUT_SECS: 20                  # Read timeout in seconds.
-   S3_ENABLE_STATS: true                                # Enable the Stats feature. Default is false.
+   S3_ENABLE_STATS: false                               # Enable the Stats feature. Default is false.
    S3_STATSD_IP_ADDR: 127.0.0.1                         # StatsD server IP address
    S3_STATSD_PORT: 8125                                 # StatsD server port
    S3_STATSD_MAX_SEND_RETRY: 3                          # Limit the user requested retry count. A retry is attempted in case message delivery to StatsD server fails.
@@ -73,13 +73,13 @@
    S3_UNIT_SIZES_FOR_MEMORY_POOL: [4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576] # Memory pool will be created for each of these unit_size with following properties
    S3_CLOVIS_READ_POOL_INITIAL_BUFFER_COUNT: 10        # 10 blocks, the initial pool size = multiple of S3_CLOVIS_UNIT_SIZE
    S3_CLOVIS_READ_POOL_EXPANDABLE_COUNT: 50            # 50 blocks, pool's expandable size, multiple of S3_CLOVIS_UNIT_SIZE
-   S3_CLOVIS_READ_POOL_MAX_THRESHOLD: 1048576000        # 1GB, The maximum memory threshold for the pool, multiple of S3_CLOVIS_UNIT_SIZE
+   S3_CLOVIS_READ_POOL_MAX_THRESHOLD: 524288000        # 500 MB, The maximum memory threshold for the pool, multiple of S3_CLOVIS_UNIT_SIZE
    S3_CLOVIS_OPERATION_WAIT_PERIOD: 90                  # 90 s, Maximum wait duration for sync clovis operations.
 S3_THIRDPARTY_CONFIG:
    S3_LIBEVENT_POOL_BUFFER_SIZE: 16384                  # Pool buffer size, in case of S3_CLOVIS_UNIT_SIZE of size 1MB, it is recommended to have S3_LIBEVENT_POOL_BUFFER_SIZE of size 16384
    S3_LIBEVENT_MAX_READ_SIZE: 16384                     # Maximum read in a single read operation, as per libevent documentation in code, user should not try to read more than this value
-   S3_LIBEVENT_POOL_INITIAL_SIZE: 10737418240           # 10GB, the initial pool size, multiple of S3_CLOVIS_UNIT_SIZE
+   S3_LIBEVENT_POOL_INITIAL_SIZE: 209715200             # Two Hundred 1mb blocks, the initial pool size, multiple of S3_CLOVIS_UNIT_SIZE
    S3_LIBEVENT_POOL_EXPANDABLE_SIZE: 104857600          # 100mb, pool's expandable size, multiple of S3_CLOVIS_UNIT_SIZE
-   S3_LIBEVENT_POOL_MAX_THRESHOLD: 10737418240          # 10GB, The maximum memory threshold for the pool, multiple of S3_CLOVIS_UNIT_SIZE
-   S3_LIBEVENT_POOL_RESERVE_SIZE: 1073741824            # Deny PUT request if mempool free space is less than the mentioned size in bytes
+   S3_LIBEVENT_POOL_MAX_THRESHOLD: 524288000            # 500 MB, The maximum memory threshold for the pool, multiple of S3_CLOVIS_UNIT_SIZE
+   S3_LIBEVENT_POOL_RESERVE_SIZE: 1048576               # Deny PUT request if mempool free space is less than the mentioned size in bytes
    S3_LIBEVENT_POOL_RESERVE_PERCENT: 5                  # Deny PUT request if mempool free space in percent is less than the mentioned percent
EOF
} | sudo patch --forward /opt/seagate/s3/conf/s3config.yaml
